[k8s-1.16-kernel-4.9](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.16-kernel-4.9/630/) hit #18060, which blocked the whole run.

[test-runtime](https://jenkins.cilium.io/job/Cilium-PR-Runtime-net-next/1085/) Hit two peculiar problems:

(1) [RuntimePolicies TestsEgressToHost Tests egress with CIDR+L4 policy to external https service](https://jenkins.cilium.io/job/Cilium-PR-Runtime-net-next/1085/testReport/junit/(root)/Suite-runtime/RuntimePolicies_TestsEgressToHost_Tests_egress_with_CIDR_L4_policy_to_external_https_service/) failed to contact 1.1.1.1. Looking into the `monitor.log` file in the test artifacts, Cilium hands the echo requests to the stack, and the stack has a default route pointing out of the node. So Cilium seems to be doing the right thing, it just doesn't receive any response from Cloudflare. This PR doesn't change any related code, and I see no other reports in the issues of similar behaviour. Infra issue? :shrug: 

(2) [RuntimePrivilegedUnitTests Run Tests](https://jenkins.cilium.io/job/Cilium-PR-Runtime-net-next/1085/testReport/junit/(root)/Suite-runtime/RuntimePrivilegedUnitTests_Run_Tests/)  hit this:

```
	 flag provided but not defined: -check.vv
	 Usage of /tmp/go-build534159421/b001/prog_test.test:
	   -test.bench regexp
	     	run only benchmarks matching regexp
	   -test.benchmem
	     	print memory allocations for benchmarks
	   -test.benchtime d
	     	run each benchmark for duration d (default 1s)
	   -test.blockprofile file
	     	write a goroutine blocking profile to file
	   -test.blockprofilerate rate
	     	set blocking profile rate (see runtime.SetBlockProfileRate) (default 1)
	   -test.count n
	     	run tests and benchmarks n times (default 1)
	   -test.coverprofile file
	     	write a coverage profile to file
	   -test.cpu list
	     	comma-separated list of cpu counts to run each test with
	   -test.cpuprofile file
	     	write a cpu profile to file
	   -test.failfast
	     	do not start new tests after the first test failure
	   -test.list regexp
	     	list tests, examples, and benchmarks matching regexp then exit
	   -test.memprofile file
	     	write an allocation profile to file
	   -test.memprofilerate rate
	     	set memory allocation profiling rate (see runtime.MemProfileRate)
	   -test.mutexprofile string
	     	write a mutex contention profile to the named file after execution
	   -test.mutexprofilefraction int
	     	if >= 0, calls runtime.SetMutexProfileFraction() (default 1)
	   -test.outputdir dir
	     	write profiles to dir
	   -test.paniconexit0
	     	panic on call to os.Exit(0)
	   -test.parallel n
	     	run at most n tests in parallel (default 4)
	   -test.run regexp
	     	run only tests and examples matching regexp
	   -test.short
	     	run smaller test suite to save time
	   -test.shuffle string
	     	randomize the execution order of tests and benchmarks (default "off")
	   -test.testlogfile file
	     	write test action log to file (for use only by cmd/go)
	   -test.timeout d
	     	panic test binary after duration d (default 0, timeout disabled)
	   -test.trace file
	     	write an execution trace to file
	   -test.v
	     	verbose: print additional output
	 FAIL	github.com/cilium/cilium/test/test_results/RuntimePolicies_TestsEgressToHost_Tests_egress_with_CIDR+L4_policy_to_external_https_service/run/cilium/bpf/tests/prog_test	0.016s
	 FAIL
	 make: Leaving directory '/home/vagrant/go/src/github.com/cilium/cilium'
	 
Stderr:
 	 make: *** [Makefile:138: tests-privileged] Error 1
```

I don't understand why this would occur on this PR and not elsewhere ([master seems green](https://jenkins.cilium.io/job/cilium-master-runtime-kernel-net-next/)). The Makefile line is the line where we iterate through all of the test packages where there are privileged tests. The `check.vv` parameter is a flag to gocheck, so plausibly if we removed gocheck from some test recently then this could explain it, but I see no such evidence in git logs.  It's also odd that the directory for running these unit tests is under the directory for the previous failing test, could the previous failing test somehow have inadvertently triggered this failure by triggering the privileged tests from the wrong directory? :shrug: 

All other tests passed, despite previous failures :thinking: . I guess I'll just re-trigger CI again to go fishing for more failures.

--- a/test/helpers/kubectl.go
+++ b/test/helpers/kubectl.go
@@ -310,6 +310,10 @@ type Kubectl struct {
 	// ciliumOptions is a cache of the most recent configuration options
 	// used to install Cilium via CiliumInstall().
 	ciliumOptions map[string]string
+
+	// nDNSReplicas is the number of replicas for DNS pods in the cluster.
+	// Stored via kub.ScaleDownDNS(), used by kub.ScaleUpDNS().
+	nDNSReplicas int
 }
 
 // CreateKubectl initializes a Kubectl helper with the provided vmName and log
@@ -2047,11 +2051,63 @@ iteratePods:
 	}
 }
 
+func (kub *Kubectl) setDNSReplicas(nReplicas int) *CmdRes {
+	res := kub.ExecShort(fmt.Sprintf("%s get deploy -n %s -l %s -o jsonpath='{.items[*].metadata.name}'", KubectlCmd, KubeSystemNamespace, kubeDNSLabel))
+	if !res.WasSuccessful() {
+		return res
+	}
+
+	// kubectl -n kube-system patch deploy coredns --patch '{"spec": { "replicas":1}}'
+	name := res.Stdout()
+	spec := fmt.Sprintf("{\"spec\": { \"replicas\":%d}}", nReplicas)
+	return kub.ExecShort(fmt.Sprintf("%s patch deploy -n %s %s --patch '%s'", KubectlCmd, KubeSystemNamespace, name, spec))
+}
+
+// ScaleDownDNS reduces the number of pods in the cluster performing kube-dns
+// duties down to zero. May be reverted by calling ScaleUpDNS().
+func (kub *Kubectl) ScaleDownDNS() *CmdRes {
+	cmd := fmt.Sprintf("%s get deploy -n %s -l %s -o jsonpath='{.items[*].status.replicas}'", KubectlCmd, KubeSystemNamespace, kubeDNSLabel)
+	res := kub.ExecShort(cmd)
+	if !res.WasSuccessful() {
+		ginkgoext.Failf("Unable to retrieve DNS pods to scale down, command '%s': %s", res.GetCmd(), res.OutputPrettyPrint())
+		return res
+	}
+
+	n, err := strconv.Atoi(res.Stdout())
+	if err != nil {
+		ginkgoext.Failf("Failed to retrieve DNS replicas via '%s': %s", res.GetCmd(), err)
+		res.success = false
+		res.err = err
+		return res
+	}
+	kub.nDNSReplicas = n
+
+	res = kub.setDNSReplicas(0)
+	if !res.WasSuccessful() {
+		ginkgoext.Failf("Unable to scale down DNS pods, command '%s': %s", res.GetCmd(), res.OutputPrettyPrint())
+	}
+	return res
+}
+
+// ScaleUpDNS restores the number of replicas for kube-dns to the number
+// prior to calling ScaleDownDNS(). Must be called after ScaleDownDNS().
+func (kub *Kubectl) ScaleUpDNS() *CmdRes {
+	res := kub.setDNSReplicas(kub.nDNSReplicas)
+	if !res.WasSuccessful() {
+		ginkgoext.Failf("Unable to scale down DNS pods, command '%s': %s", res.GetCmd(), res.OutputPrettyPrint())
+	}
+	return res
+}
+
 // RedeployDNS deletes the kube-dns pods and does not wait for the deletion
 // to complete. Useful to ensure that the pods are recreated after datapath
 // configuration changes.
 func (kub *Kubectl) RedeployDNS() *CmdRes {
-	return kub.DeleteResource("pod", "-n "+KubeSystemNamespace+" -l "+kubeDNSLabel)
+	if res := kub.ScaleDownDNS(); !res.WasSuccessful() {
+		return res
+	}
+
+	return kub.ScaleUpDNS()
 }
 
 // RedeployKubernetesDnsIfNecessary validates if the Kubernetes DNS is

--- a/test/helpers/manifest.go
+++ b/test/helpers/manifest.go
@@ -280,6 +280,8 @@ func (m *DeploymentManager) DeleteAll() {
 // DeployCilium()
 func (m *DeploymentManager) DeleteCilium() {
 	if m.ciliumDeployed {
+		// Ensure any Cilium-managed pods are terminated first
+		m.kubectl.WaitTerminatingPods(2 * time.Minute)
 		ginkgoext.By("Deleting Cilium")
 		m.kubectl.DeleteAndWait(m.ciliumFilename, true)
 		m.kubectl.WaitTerminatingPods(2 * time.Minute)

--- a/test/k8sT/Chaos.go
+++ b/test/k8sT/Chaos.go
@@ -189,6 +189,7 @@ var _ = SkipDescribeIf(helpers.RunsOn54Kernel, "K8sChaosTest", func() {
 
 		AfterAll(func() {
 			_ = kubectl.Delete(netperfManifest)
+			ExpectAllPodsTerminated(kubectl)
 		})
 
 		AfterEach(func() {

--- a/test/k8sT/CustomCalls.go
+++ b/test/k8sT/CustomCalls.go
@@ -101,8 +101,8 @@ var _ = SkipDescribeIf(func() bool {
 		})
 
 		AfterAll(func() {
-			deploymentManager.DeleteCilium()
 			deploymentManager.DeleteAll()
+			deploymentManager.DeleteCilium()
 			kubectl.RedeployDNS()
 		})
 

--- a/test/k8sT/DatapathConfiguration.go
+++ b/test/k8sT/DatapathConfiguration.go
@@ -36,6 +36,7 @@ var _ = Describe("K8sDatapathConfig", func() {
 
 	AfterEach(func() {
 		deploymentManager.DeleteAll()
+		ExpectAllPodsTerminated(kubectl)
 	})
 
 	AfterFailed(func() {

--- a/test/k8sT/Egress.go
+++ b/test/k8sT/Egress.go
@@ -110,6 +110,7 @@ var _ = SkipDescribeIf(func() bool {
 	AfterAll(func() {
 		_ = kubectl.Delete(echoPodYAML)
 		_ = kubectl.Delete(assignIPYAML)
+		ExpectAllPodsTerminated(kubectl)
 
 		UninstallCiliumFromManifest(kubectl, ciliumFilename)
 	})

--- a/test/k8sT/Health.go
+++ b/test/k8sT/Health.go
@@ -39,11 +39,8 @@ var _ = Describe("K8sHealthTest", func() {
 			kubectl.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
 		})
 
-		AfterEach(func() {
-			ExpectAllPodsTerminated(kubectl)
-		})
-
 		AfterAll(func() {
+			ExpectAllPodsTerminated(kubectl)
 			UninstallCiliumFromManifest(kubectl, ciliumFilename)
 			kubectl.CloseSSHClient()
 		})

--- a/test/k8sT/Policies.go
+++ b/test/k8sT/Policies.go
@@ -109,10 +109,6 @@ var _ = SkipDescribeIf(func() bool {
 		DeployCiliumOptionsAndDNS(kubectl, ciliumFilename, daemonCfg)
 	})
 
-	AfterEach(func() {
-		ExpectAllPodsTerminated(kubectl)
-	})
-
 	AfterFailed(func() {
 		kubectl.CiliumReport("cilium service list", "cilium endpoint list")
 	})
@@ -218,6 +214,7 @@ var _ = SkipDescribeIf(func() bool {
 		AfterAll(func() {
 			kubectl.NamespaceDelete(namespaceForTest)
 			kubectl.Delete(demoPath)
+			ExpectAllPodsTerminated(kubectl)
 		})
 
 		BeforeEach(func() {
@@ -1500,6 +1497,7 @@ var _ = SkipDescribeIf(func() bool {
 					// Remove echoserver pods from host namespace.
 					echoPodPath := helpers.ManifestGet(kubectl.BasePath(), "echoserver-cilium-hostnetns.yaml")
 					kubectl.Delete(echoPodPath).ExpectSuccess("Cannot remove echoserver application")
+					ExpectAllPodsTerminated(kubectl)
 				})
 
 				It("Connectivity to hostns is blocked after denying ingress", func() {
@@ -1759,6 +1757,7 @@ var _ = SkipDescribeIf(func() bool {
 
 			AfterEach(func() {
 				kubectl.Delete(connectivityCheckYml)
+				ExpectAllPodsTerminated(kubectl)
 			})
 
 			It("using connectivity-check to check datapath", func() {
@@ -1969,6 +1968,7 @@ var _ = SkipDescribeIf(func() bool {
 			_ = kubectl.Delete(demoManifest)
 			_ = kubectl.Delete(cnpSecondNS)
 			_ = kubectl.NamespaceDelete(secondNS)
+			ExpectAllPodsTerminated(kubectl)
 		})
 
 		It("Tests the same Policy in different namespaces", func() {
@@ -2179,6 +2179,7 @@ var _ = SkipDescribeIf(func() bool {
 			_ = kubectl.Delete(demoManifestNS2)
 			_ = kubectl.NamespaceDelete(firstNS)
 			_ = kubectl.NamespaceDelete(secondNS)
+			ExpectAllPodsTerminated(kubectl)
 		})
 
 		It("Test clusterwide connectivity with policies", func() {

--- a/test/k8sT/Services.go
+++ b/test/k8sT/Services.go
@@ -134,10 +134,6 @@ var _ = SkipDescribeIf(helpers.RunsOn54Kernel, "K8sServicesTest", func() {
 		kubectl.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
 	})
 
-	AfterEach(func() {
-		ExpectAllPodsTerminated(kubectl)
-	})
-
 	AfterAll(func() {
 		UninstallCiliumFromManifest(kubectl, ciliumFilename)
 		kubectl.CloseSSHClient()
@@ -232,6 +228,7 @@ var _ = SkipDescribeIf(helpers.RunsOn54Kernel, "K8sServicesTest", func() {
 					_ = kubectl.Delete(echoSVCYAMLDualStack)
 				}
 			}
+			ExpectAllPodsTerminated(kubectl)
 		})
 
 		SkipItIf(helpers.RunsWithKubeProxyReplacement, "Checks service on same node", func() {
@@ -1046,6 +1043,7 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 
 						AfterAll(func() {
 							kubectl.Delete(echoYAML)
+							ExpectAllPodsTerminated(kubectl)
 						})
 
 						It("Tests NodePort", func() {
@@ -1186,6 +1184,7 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 
 						AfterAll(func() {
 							_ = kubectl.Delete(echoYAML)
+							ExpectAllPodsTerminated(kubectl)
 						})
 
 						It("Tests NodePort", func() {
@@ -1369,6 +1368,7 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 							for _, node := range []string{ni.k8s1NodeName, ni.k8s2NodeName, ni.outsideNodeName} {
 								kubectl.ExecInHostNetNS(context.TODO(), node, "ip l del wg0")
 							}
+							ExpectAllPodsTerminated(kubectl)
 						})
 
 						It("Tests NodePort BPF", func() {
@@ -1659,6 +1659,7 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 		AfterEach(func() {
 			wg.Wait()
 			_ = kubectl.Delete(gracefulTermYAML)
+			ExpectAllPodsTerminated(kubectl)
 		})
 
 		It("Checks client terminates gracefully on service endpoint deletion", func() {

--- a/test/k8sT/Verifier.go
+++ b/test/k8sT/Verifier.go
@@ -140,6 +140,7 @@ var _ = Describe("K8sVerifier", func() {
 
 	AfterAll(func() {
 		kubectl.DeleteResource("pod", podName)
+		ExpectAllPodsTerminated(kubectl)
 	})
 
 	It("Runs the kernel verifier against Cilium's BPF datapath", func() {

--- a/test/k8sT/bookinfo.go
+++ b/test/k8sT/bookinfo.go
@@ -34,10 +34,6 @@ var _ = SkipDescribeIf(func() bool {
 		kubectl.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
 	})
 
-	AfterEach(func() {
-		ExpectAllPodsTerminated(kubectl)
-	})
-
 	AfterAll(func() {
 		UninstallCiliumFromManifest(kubectl, ciliumFilename)
 		kubectl.CloseSSHClient()
@@ -84,6 +80,7 @@ var _ = SkipDescribeIf(func() bool {
 				// Explicitly do not check result to avoid having assertions in AfterAll.
 				_ = kubectl.Delete(resourcePath)
 			}
+			ExpectAllPodsTerminated(kubectl)
 		})
 
 		It("Tests bookinfo demo", func() {

--- a/test/k8sT/hubble.go
+++ b/test/k8sT/hubble.go
@@ -164,13 +164,10 @@ var _ = Describe("K8sHubbleTest", func() {
 			kubectl.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
 		})
 
-		AfterEach(func() {
-			ExpectAllPodsTerminated(kubectl)
-		})
-
 		AfterAll(func() {
 			kubectl.Delete(demoPath)
 			kubectl.NamespaceDelete(namespaceForTest)
+			ExpectAllPodsTerminated(kubectl)
 
 			kubectl.DeleteHubbleRelay(hubbleRelayNamespace)
 			UninstallCiliumFromManifest(kubectl, ciliumFilename)

--- a/test/k8sT/istio.go
+++ b/test/k8sT/istio.go
@@ -129,6 +129,7 @@ var _ = SkipDescribeIf(helpers.RunsOn54Kernel, "K8sIstioTest", func() {
 
 		By("Deleting the istio-system namespace")
 		_ = kubectl.NamespaceDelete(istioSystemNamespace)
+		ExpectAllPodsTerminated(kubectl)
 
 		UninstallCiliumFromManifest(kubectl, ciliumFilename)
 		kubectl.CloseSSHClient()

